language/sygus.tex

```diff
@@ -1895,17 +1892,80 @@ \subsection{Satisfying Semantic Specifications}
+$(\tau_1 \times \ldots \times \tau_m)$ if
+it has the following behavior.
+Let \textit{textfile.oracle}
+be a file containing the text
+\[
+\paren{v_1 \ldots v_n}
+\]
+where $v_1, \ldots, v_n$ are values\footnote{
+A \emph{value} of type $T$ 
+coincides with a term generated by the SyGuS grammar term
+$\paren{\constantkwd\mbox{ }T}$.
+For example, the values of type ${\tt Int}$
+have syntax ${\tt N}$ or ${\tt (- N)}$ where ${\tt N}$ is a numeral.
+The values for function types include all closed lambda terms.
+} of sorts $\sigma_1, \ldots, \sigma_n$
+written using the syntax for values described in this document.
+Let ${\tt oracle}$ be the name of the binary corresponding to the oracle.
+When the oracle is executed with the command
+${\tt oracle textfile.oracle}$,
+it returns the text on the standard output channel
+\[
+\paren{w_{1} \ldots w_{m}}
+\]
+where $w_1, \ldots, w_n$ are values of sorts $\tau_1, \ldots, \tau_m$
+written using the syntax for values described in this document.
+\qed
 \end{definition}
-The SyGuS solver makes queries to an oracle via a text file with the extension \textit{.oracle}. The oracle is called by the solver with the command \textit{oracle-name textfile.oracle}.
-The text file should contain the text $\paren{v_1 \ldots v_n}$ specifying values for each argument to the oracle, where $v_1 \ldots v_n$ are written using the same syntax for constant terms described in this document. The oracle will return the text $\paren{w_{1} \ldots w_{m}}$ specifying values for the return values of the oracle on the standard output channel.\\
-
 
+In other words, the synthesis solver and the external binary
+communicate with one another using a text interface, where the input text contains
+a tuple of values corresponding to input to the binary, and the output
+from the binary is a tuple of terms indicating the output of the binary,
+which is printed on the standard output channel.
+It is expected that the synthesis solver parses these values to
+ascertain the result of the query.
+
+It is important to note that the text file
+${\tt textfile.oracle}$ in the definition above is expected to take
+a tuple of values only.
+In some cases, 
+the syntax for those values may involve symbols that are user-defined.
+As an example, consider an oracle
+implementing an interface whose input takes a user-defined datatype.
+The definition of that datatype is \emph{not} provided as a preamble
+to the given tuple of values.
+This means that an external binary is responsible for being able to
+parse the given text file, even when it uses symbols that
+are not self-contained in that text file. 
+This is a reasonable assumption,
+considering that the user is providing both the oracle specification
+and the binary.
+
+The oracle accepts the name of the text file as the first command line
+argument. An oracle may optionally take additional command line
+arguments which are not standardized in this document.
```

Comment:
may it? how does the user provide that to the synthesis solver? 

I was going to just force the user to wrap any oracles that need additional args in a script

Reply:
So I am imagining that a synthesis solver / oracle could in theory be hardcoded to each other, in which case the solver may want to provide additional command line arguments.  This is somewhat analogous to how SMT-LIB/SyGuS formalizes `(set-option :name ...)` but doesn't insist on what "name" is.

This is a bit of a minor issue, I could perhaps omit this.

Reply:
I think including that option raises more questions than it answers at this point: if we say it can take additional arguments, i think we should also explain how. I'd either omit, or elabprate on the `set-option` idea